File: app/src/main/java/io/github/marcocipriani01/telescopetouch/renderer/util/TexCoordBuffer.java
/*
 * Copyright 2021 Marco Cipriani (@marcocipriani01)
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package io.github.marcocipriani01.telescopetouch.renderer.util;

import android.util.Log;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.IntBuffer;

import javax.microedition.khronos.opengles.GL10;
import javax.microedition.khronos.opengles.GL11;

import io.github.marcocipriani01.telescopetouch.maths.MathsUtils;

public class TexCoordBuffer {

    private final GLBuffer gLBuffer = new GLBuffer(GL11.GL_ARRAY_BUFFER);
    private IntBuffer texCoordBuffer = null;
    private int numVertices = 0;
    private boolean useVBO = false;

    public TexCoordBuffer(int numVertices) {
        reset(numVertices);
    }

    public TexCoordBuffer(boolean useVBO) {
        numVertices = 0;
        this.useVBO = useVBO;
    }

    public int size() {
        return numVertices;
    }

    public void reset(int numVertices) {
        if (numVertices < 0) {
            Log.e("TexCoordBuffer", "reset attempting to set numVertices to " + numVertices);
            numVertices = 0;
        }
        this.numVertices = numVertices;
        regenerateBuffer();
    }

    // Call this when we have to re-create the surface and reloading all OpenGL resources.
    public void reload() {
        gLBuffer.reload();
    }

    public void addTexCoords(float u, float v) {
        texCoordBuffer.put(MathsUtils.floatToFixedPoint(u));
        texCoordBuffer.put(MathsUtils.floatToFixedPoint(v));
    }

    public void set(GL10 gl) {
        if (numVertices == 0) {
            return;
        }
        texCoordBuffer.position(0);

        if (useVBO && GLBuffer.canUseVBO()) {
            GL11 gl11 = (GL11) gl;
            gLBuffer.bind(gl11, texCoordBuffer, 4 * texCoordBuffer.capacity());
            gl11.glTexCoordPointer(2, GL10.GL_FIXED, 0, 0);
        } else {
            gl.glTexCoordPointer(2, GL10.GL_FIXED, 0, texCoordBuffer);
        }
    }

    private void regenerateBuffer() {
        if (numVertices == 0) {
            return;
        }

        ByteBuffer bb = ByteBuffer.allocateDirect(4 * 2 * numVertices);
        bb.order(ByteOrder.nativeOrder());
        IntBuffer ib = bb.asIntBuffer();
        ib.position(0);
        texCoordBuffer = ib;
    }
}

File: app/src/main/java/com/google/android/stardroid/renderer/util/TexCoordBuffer.kt
// Copyright 2008 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package com.google.android.stardroid.renderer.util

import android.util.Log
import com.google.android.stardroid.util.FixedPoint.floatToFixedPoint
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.IntBuffer
import javax.microedition.khronos.opengles.GL10
import javax.microedition.khronos.opengles.GL11

class TexCoordBuffer {
    constructor(numVertices: Int) {
        reset(numVertices)
    }

    // Creates an empty buffer.  Must call reset() before adding vertices.
    constructor() {
        mNumVertices = 0
    }

    constructor(useVBO: Boolean) {
        mNumVertices = 0
        mUseVBO = useVBO
    }

    fun size(): Int {
        return mNumVertices
    }

    fun reset(numVertices: Int) {
        var numVertices = numVertices
        if (numVertices < 0) {
            Log.e("TexCoordBuffer", "reset attempting to set numVertices to $numVertices")
            numVertices = 0
        }
        mNumVertices = numVertices
        regenerateBuffer()
    }

    // Call this when we have to re-create the surface and reloading all OpenGL resources.
    fun reload() {
        mGLBuffer.reload()
    }

    fun addTexCoords(u: Float, v: Float) {
        mTexCoordBuffer!!.put(floatToFixedPoint(u))
        mTexCoordBuffer!!.put(floatToFixedPoint(v))
    }

    fun set(gl: GL10) {
        if (mNumVertices == 0) {
            return
        }
        mTexCoordBuffer!!.position(0)
        if (mUseVBO && GLBuffer.canUseVBO()) {
            val gl11 = gl as GL11
            mGLBuffer.bind(gl11, mTexCoordBuffer!!, 4 * mTexCoordBuffer!!.capacity())
            gl11.glTexCoordPointer(2, GL10.GL_FIXED, 0, 0)
        } else {
            gl.glTexCoordPointer(2, GL10.GL_FIXED, 0, mTexCoordBuffer)
        }
    }

    private fun regenerateBuffer() {
        if (mNumVertices == 0) {
            return
        }
        val bb = ByteBuffer.allocateDirect(4 * 2 * mNumVertices)
        bb.order(ByteOrder.nativeOrder())
        val ib = bb.asIntBuffer()
        ib.position(0)
        mTexCoordBuffer = ib
    }

    private var mTexCoordBuffer: IntBuffer? = null
    private var mNumVertices = 0
    private val mGLBuffer = GLBuffer(GL11.GL_ARRAY_BUFFER)
    private var mUseVBO = false
}